### PR TITLE
Dev Rollup

### LINKED ISSUE #712: Dialogs switch the stack to an incorrect window
(1) Issue/Bug Description
When using window stacking, popos changes active stack window when a IntelliJ IDEA based product in a stack application launches a dialog (such as Settings)

(2) Steps to reproduce (if you know):

Create a popos stack
Add a Jetbrain's IntelliJ IDEA (could also be a variant such as WebStorm, Rider, CLion, PyCharm, PhpStorm or RubyMine, DataGrip, AppCode or Android Studio)
Add another app such as Firefox.  
Do an action that launches a dialog from  the IntelliJ such as opening settings (File -> Settings) popos shell needlessly switches the stack to the wrong window. 

(3) Expected behavior:  No switching of my stack

(4) Distribution (run cat /etc/os-release): Pop!_OS 20.10

(5) Gnome Shell version: 3.38.1

(6) Pop Shell version (run apt policy pop-shell or provide the latest commit if building locally):

1.0.0~1604688242~20.10~d1d2b0e

(7) Where was Pop Shell installed from: apt-get install popos-desktop

I found this https://github.com/pop-os/shell/issues/407 but it doesn't explain why the stack is changing active window, unless it's expecting the dialog to open as part of the stack and trying to focus it but it isn't there?? 

Perhaps it should only refocus the stack if a window is actually added to the stack, as the modal appears correctly in all other manners?

### LINKED ISSUE #856: Tile Windows Toggle not updating when using hotkeys
**(1) Issue/Bug Description:**

When using `super+y` the visual toggle (Check button) only updates when switching from titling to floating but not the reverse. 

See screenshot: 

Screenshot from 2021-01-27 10-12-51

**(2) Steps to reproduce (if you know):**

Turn on Tiling using hotkey and check the visual status (issue appears)
Turn off Tiling using hotkey, turn on tiling using UI 
Turn off Tiling using hotkey, check tiling using UI (it has updated)
Toggle only using hotkeys, checking UI. (no updates occur)

**(3) Expected behavior:**
Visual toggle should be in sync with tiling status.

**(4) Distribution (run `cat /etc/os-release`):**
 
NAME="Pop!_OS"
VERSION="20.10"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 20.10"
VERSION_ID="20.10"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=groovy
UBUNTU_CODENAME=groovy
LOGO=distributor-logo-pop-os

**(5) Gnome Shell version:**

```
gnome-shell:
  Installed: 3.38.2-1pop1~1610298831~20.10~41092e5
  Candidate: 3.38.2-1pop1~1610298831~20.10~41092e5
  Version table:
     3.38.2-1ubuntu1~20.10.1 500
        500 http://us.archive.ubuntu.com/ubuntu groovy-updates/main amd64 Packages
 *** 3.38.2-1pop1~1610298831~20.10~41092e5 1001
       1001 http://ppa.launchpad.net/system76/pop/ubuntu groovy/main amd64 Packages
        100 /var/lib/dpkg/status
     3.38.1-1ubuntu1 500
        500 http://us.archive.ubuntu.com/ubuntu groovy/main amd64 Packages
```

**(6) Pop Shell version (run `apt policy pop-shell` or provide the latest commit if building locally):**

```
pop-shell:
  Installed: 1.1.0~1611709512~20.10~673b87a
  Candidate: 1.1.0~1611709512~20.10~673b87a
  Version table:
 *** 1.1.0~1611709512~20.10~673b87a 1001
       1001 http://ppa.launchpad.net/system76/pop/ubuntu groovy/main amd64 Packages
       1001 http://ppa.launchpad.net/system76/pop/ubuntu groovy/main i386 Packages
        100 /var/lib/dpkg/status

```

**(7) Where was Pop Shell installed from:**

Default install on Pop!_OS including all official updates from repositories.

**(8) Monitor Setup (2 x 1080p, 4K, Primary(Horizontal), Secondary(Vertical), etc):**

N/A

**(9) Other Installed/Enabled Extensions:**

None

**(10) Other Notes:**

### LINKED ISSUE #877: Obsidian dialog window elements not clickable when tilling windows is on
**(1) Issue/Bug Description:**

When using [Obsidian](https://obsidian.md/) app and pop-shell with "Tilling Windows" enabled, some Obsidian windows are not clickable. See

example

I am clicking on all the buttons, but it´s doing nothing.

If I disable "Tilling Window" in pop-shell, it works. That´s why I think it can be an issue on pop-shell rather than Obsidian. 
I know Obsidian is built with Electron and the particular window that is not working should behave like a dialog window,

**(2) Steps to reproduce (if you know):**

Install  [Obsidian](https://obsidian.md/) latest version (I have installed the one from flathub).
With "Tilling windows" option active, open Obsidian app and click on "Open Another Vault" (the first icon on the bottom in the left bar). It will open a new window where buttons are not clickable.

**(3) Expected behavior:**

The window elements should be clickable.

**(4) Distribution (run `cat /etc/os-release`):**
 
```
NAME=Fedora
VERSION="33 (Workstation Edition)"
ID=fedora
VERSION_ID=33
VERSION_CODENAME=""
PLATFORM_ID="platform:f33"
PRETTY_NAME="Fedora 33 (Workstation Edition)"
ANSI_COLOR="0;38;2;60;110;180"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:33"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f33/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=33
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=33
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Workstation Edition"
VARIANT_ID=workstation
```

**(5) Gnome Shell version:**

GNOME Shell 3.38.3

**(6) Pop Shell version (run `apt policy pop-shell` or provide the latest commit if building locally):**

```gnome-shell-extension-pop-shell-1.2.0-1.fc33.noarch```

**(7) Where was Pop Shell installed from:**

Fedora package.

**(8) Monitor Setup (2 x 1080p, 4K, Primary(Horizontal), Secondary(Vertical), etc):**

2 Horzontal Monitors (1080p)

**(9) Other Installed/Enabled Extensions:**

```
windowIsReady_Remover@nunofarruca@gmail.com
dash-to-dock@micxgx.gmail.com
CoverflowAltTab@palatis.blogspot.com
emoji-selector@maestroschan.fr
user-theme@gnome-shell-extensions.gcampax.github.com
gsconnect@andyholmes.github.io
pop-shell@system76.com
appindicatorsupport@rgcjonas.gmail.com
jetbrains-search-provider@swsnr.de
mprisindicatorbutton@JasonLG1979.github.io
status-area-horizontal-spacing@mathematical.coffee.gmail.com
Vitals@CoreCoding.com
launch-new-instance@gnome-shell-extensions.gcampax.github.com
trayIconsReloaded@selfmade.pl
github.notifications@alexandre.dufournet.gmail.com
auto-move-windows@gnome-shell-extensions.gcampax.github.com
caffeine@patapon.info
timepp@zagortenay333
vscode-search-provider@jomik.org
clipboard-indicator@tudmotu.com
gitlab-extension@infinicode.de
proxyprofiles@massamany.github.com
easy_docker_containers@red.software.systems
remove-dropdown-arrows@mpdeimos.com
always-indicator@martin.zurowietz.de
cast-to-tv@rafostar.github.com
background-logo@fedorahosted.org
```

**(10) Other Notes:**

Detail of the problematic window using xprop:

```
_NET_WM_STATE(ATOM) = 
WM_STATE(WM_STATE):
		window state: Normal
		icon window: 0x0
_NET_WM_DESKTOP(CARDINAL) = 2
_GTK_EDGE_CONSTRAINTS(CARDINAL) = 170
_NET_WM_ALLOWED_ACTIONS(ATOM) = _NET_WM_ACTION_MOVE, _NET_WM_ACTION_MINIMIZE, _NET_WM_ACTION_CHANGE_DESKTOP, _NET_WM_ACTION_CLOSE, _NET_WM_ACTION_ABOVE, _NET_WM_ACTION_BELOW
_NET_WM_USER_TIME(CARDINAL) = 19229902
WM_NAME(UTF8_STRING) = "Obsidian"
_NET_WM_NAME(UTF8_STRING) = "Obsidian"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
		program specified location: 560, 228
		program specified minimum size: 800 by 650
		program specified maximum size: 800 by 650
_MOTIF_WM_HINTS(_MOTIF_WM_HINTS) = 0x2, 0x0, 0x0, 0x0, 0x0
XdndAware(ATOM) = BITMAP
_NET_WM_BYPASS_COMPOSITOR(CARDINAL) = 2
_GTK_HIDE_TITLEBAR_WHEN_MAXIMIZED(CARDINAL) = 1
WM_WINDOW_ROLE(STRING) = "browser-window"
WM_CLASS(STRING) = "obsidian", "obsidian"
_NET_WM_WINDOW_TYPE(ATOM) = _NET_WM_WINDOW_TYPE_NORMAL
_NET_WM_PID(CARDINAL) = 4
WM_CLIENT_MACHINE(STRING) = "bruno-zenbook"
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING, _NET_WM_SYNC_REQUEST
```

### LINKED ISSUE #889: When dragging window over adjacent stack, behavior is inconsistent and doesn't match overlay
**(1) Issue/Bug Description:** When dragging a window over a stack next to the window, the overlay's behavior is inconsistent and sometimes unexpected:
- Sometimes, the overlay presents two possible options for swapping positions and one for stacking. (From left to right, the overlay would swap, swap, or add to stack.)
- Other times, the overlay presents one possible option for swapping positions and two options for stacking. (From left to right, the overlay would swap, add to stack, or add to stack.)
There's no visual difference between these two cases, it's just the behavior that changes.

**(2) Steps to reproduce (if you know):**
1. Open two windows next to each other.
2. Create a stack from the leftmost window.
3. Drag the right window over the stack.

**(3) Expected behavior:** One of these two schemes would be more consistent:
- An overlay covering the entire stack swaps positions; an overlay covering both the stack and the dragged window's starting position adds the window to the stack; an overlay covering the dragged window's starting position doesn't move anything. (It seems like this one's more in line with how the overlay currently behaves.)
- An overlay covering the left half of the stack swaps positions; an overlay covering the entire stack adds the window to the stack; an overlay covering the right half of the stack doesn't move anything.

Whichever of these is deemed correct, it should work the same way every time.

**(4) Distribution (run `cat /etc/os-release`):** Pop!_OS 20.10.

**(5) Gnome Shell version:** `3.38.3-2ubuntu0.20.10.2pop1~1614116211~20.10~0176ca0`

**(6) Pop Shell version (run `apt policy pop-shell` or provide the latest commit if building locally):** Occurs on:
- `master` (`6b2287f`)
- `unstack` (`fe2bb4a`)
- pre-`mousefix` (`e00fb3f`)

**(7) Where was Pop Shell installed from:** Staging repo or built locally.

**(8) Monitor Setup (2 x 1080p, 4K, Primary(Horizontal), Secondary(Vertical), etc):** 1x 1080p.

**(9) Other Installed/Enabled Extensions:** Pop!_OS defaults.

**(10) Other Notes:** Video: https://www.dropbox.com/s/l9pbnmpgc8fcpco/shell-pr888-screencast1.mkv?dl=0

### LINKED ISSUE #906: Active Hint Color Picker Crash
**(1) Issue/Bug Description:**
If the color picker dialog for the active hint color is opened and the user hits "Cancel" instead of "Accept", the dialog hangs for a moment and then closes followed by a crash report dialog from the system. This happened multiple times in a row, but once I hit accept (without even changing the value) the dialog no longer crashes. This may only be a problem with a fresh install since I am unable to reproduce after clicking accept in the dialog.

**(2) Steps to reproduce (if you know):**
1) Fresh install of pop-shell
1) Open color picker
1) Cancel out of dialog rather than accept.
**Crash**

**(3) Expected behavior:**
Dialog exits cleanly and does not crash.

**(4) Distribution (run `cat /etc/os-release`):**
 NAME="Ubuntu"
VERSION="20.04.2 LTS (Focal Fossa)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 20.04.2 LTS"
VERSION_ID="20.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=focal
UBUNTU_CODENAME=focal

**(5) Gnome Shell version:**
GNOME Shell 3.36.4

**(6) Pop Shell version (run `apt policy pop-shell` or provide the latest commit if building locally):**

I did a local build, but I built it in my tmp directory. I built it on February 9th from master. So the version should be on [this commit](https://github.com/pop-os/shell/commit/190ad70fe9e7ebea928892443e3787f86d1dddb7).

**(7) Where was Pop Shell installed from:**
The git repo

**(8) Monitor Setup (2 x 1080p, 4K, Primary(Horizontal), Secondary(Vertical), etc):**
2 x 1080p, Secondary monitor (TV) is vertical, primary monitor is horizontal, running on laptop with built-in display closed and not active.

**(9) Other Installed/Enabled Extensions:**
Most are unrelated:
Desktop Icons
Lock Keys
OpenWeather
Sound Input & Output Device Chooser
System76 Power
Ubuntu AppIndicators (disabled)
Ubuntu Dock (disabled)

**(10) Other Notes:**

### PR DESCRIPTION
**Pull request by mmstick**
Closes #712 
Closes #856 
Closes #877 
Closes #889 
Closes #906 

- Also fixes a maximized stacked window causing the tab to remain invisible if alt+tabbed to an sibling window

### CODE REVIEW

**Comment by jacobgkau** [comment]
On this branch, if I start with a layout including stacked and non-stacked windows:

Screenshot from 2021-03-22 09-47-59

Then move two windows to a new workspace using the keyboard, one from the stack, one not from the stack (just making sure the stack still exists):

Screenshot from 2021-03-22 09-48-07

Then drag one of those two windows to a new in-between workspace, and press Super to return to the workspace I'm currently on, then I see stacking tabs from the first workspace when I shouldn't:

Screenshot from 2021-03-22 09-48-15

I've also had windows remain visible behind other windows after I drag them to a new workspace (without switching to that new workspace.)

**Comment by jacobgkau** [comment]
After the latest changes, the active hint noticeably lags behind the cursor when switching focus with the keyboard. Before:

pr910-before

After:

pr910-after

When switching focus with the mouse, the new hint isn't delayed, but the old hint going away is delayed, which means both are briefly displayed at the same time:

pr910-mouse

**Review by jacobgkau — APPROVED** [review]
This no longer fixes #712 after the last revert. Confirmed this fixes #856, #877, #889, and #906. Regression testing passed:
[shell-pr910-groovy.txt](https://github.com/pop-os/shell/files/6281195/shell-pr910-groovy.txt)
[shell-pr910-focal.txt](https://github.com/pop-os/shell/files/6281258/shell-pr910-focal.txt)

File: src/extension.ts [hunk 1]
```diff
@@ -697,40 +691,34 @@ export class Ext extends Ecs.System<ExtEvent> {
...
-                }
+            if (null !== win.stack) {
+                ext?.auto_tiler?.forest.stacks.get(win.stack)?.activate(win.entity)
             }
         }
 
         this.show_border_on_focused();
 
-        if (this.auto_tiler && this.prev_focused !== null && win.is_tilable(this)) {
+        if (this.auto_tiler && this.prev_focused !== null) {
```
Comment (yasserkaddour) [inline src/extension.ts:705]: This line is responsible for https://github.com/pop-os/shell/issues/942, moving the mouse in VLC displays the full-screen controller windows that trigger `unmake_fullscreen()` whitout the `win.is_tilable(this)` check.

**Pull request merged**